README.md
# stack_grouping_list

Stack grouping list inpired by iOS notification group.

[![Pub](https://img.shields.io/pub/v/stack_grouping_list)](https://pub.dev/packages/stack_grouping_list)

![Screenshot](https://raw.githubusercontent.com/chittapon/stack_grouping_list/master/images/demo.gif)

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  stack_grouping_list: ^0.1.0
```

## Basic Usage
Create grouping container widget from `StackGroupingContainer` required `itemCount` and `itemBuilder` parameter.
```dart
    //For this example, we create single group.
    return StackGroupingContainer(
      itemBuilder: (BuildContext context, int index, bool isExpanded){
        List<BoxShadow> boxShadow = [BoxShadow(blurRadius: 6, color: Colors.black.withOpacity(0.04), offset: Offset(0, 3),)];
        // Prevent shadow too dark
        if (!isExpanded){
          if (index > 3 ){
            boxShadow = null;
          }
        }
        return Container(
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(10),
            color: Colors.white,
            boxShadow: boxShadow,
          ),
          padding: EdgeInsets.all(10),
          child: Column(
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.start,
                children: [
                  Icon(Icons.notifications),
                  SizedBox(width: 10,),
                  Text('MESSENGER'),
                  Spacer(),
                  Text('1m ago'),
                ],
              ),
              SizedBox(height: 4,),
              Expanded(
                child: Align(
                  alignment: Alignment.centerLeft,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('John Wick',
                      ),
                      SizedBox(height: 4,),
                      Text('Everything\'s got a price.',
                      ),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      itemCount: 3,
      headerBuilder: (BuildContext context, bool isExpanded, VoidCallback onDismiss){
        return Padding(
          padding: const EdgeInsets.only(bottom: 12),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Align(
                alignment: Alignment.bottomLeft,
                child: Text('Messenger',
                  style: TextStyle(
                    fontSize: 24,
                  ),
                ),
              ),
              Spacer(),
              Align(
                alignment: Alignment.bottomRight,
                child: GestureDetector(
                  onTap: onDismiss,
                  child: Row(
                    children: [
                      Text('Show less',
                        style: TextStyle(
                          fontSize: 17,
                        ),
                      ),
                      Icon(Icons.keyboard_arrow_up),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      groupingItemCount: 3,
      itemHeight: 90,
      headerHeight: 46,
      expandItemPadding: 15,
      onSelectItem: (index){
        print("select item $index");
      },
    );
```

## Constructor

| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| itemCount | null  | The number of item in group. |
| itemBuilder | null | Customization of item widget. |
| headerBuilder | null |  Customization of header widget. |
| isExpanded | false | Initial expand group, set true to initial expand. |
| groupingItemCount | 3  | Number of display stack cards. |
| itemHeight | 120  | Because the item widget is in the stack widget tree so we have to define height of item widget. |
| groupingItemPadding | 10.0  | Padding for each item in state grouping. |
| expandItemPadding | 10.0 | Padding for each item in state expand. |
| onSelectItem | null | Callback function when selecting item. |
| headerHeight | 46 | Height of header widget. |
| animatedSizeDuration | 350 ms | The duration of sizing animation. |
| animatedGroupDuration | 300 ms | The duration of positioned item animation. |


CHANGELOG.md
## [0.1.0]
- Initial flutter package.


test/stack_grouping_list_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:stack_grouping_list/stack_grouping_list.dart';

void main() {
}


lib/source/stack_grouping_content.dart
import 'package:flutter/material.dart';

class StackGroupingContent extends StatefulWidget {

  bool isExpanded = false;
  double itemHeight;
  int groupingItemCount;
  double groupingItemPadding;
  double expandItemPadding;
  Duration animatedGroupDuration;
  final Widget Function(BuildContext context, int index, bool isExpanded) itemBuilder;
  final Function(bool isExpanded) onContentTrigger;
  final Function(int index) onSelectItem;
  final int itemCount;

  StackGroupingContent({
    this.isExpanded,
    this.itemHeight,
    this.groupingItemCount,
    this.groupingItemPadding,
    this.expandItemPadding,
    this.animatedGroupDuration,
    this.itemBuilder,
    this.itemCount,
    this.onContentTrigger,
    this.onSelectItem});

  @override
  _StackGroupingContentState createState() => _StackGroupingContentState();
}

class _StackGroupingContentState extends State<StackGroupingContent> {

  @override
  Widget build(BuildContext context) {

    double topPosition = 0;
    double leftPosition = 0;
    double rightPosition = 0;

    return GestureDetector(
      onTap: widget.itemCount > 1 ? _expandContent : null,
      child: Stack(

        fit: StackFit.expand,
        children: List.generate(widget.itemCount, (index){

          if (widget.isExpanded){
            if (index > 0){
              topPosition += widget.itemHeight;
              topPosition += widget.expandItemPadding;
            }
          }else {
            if (index > 0 && index < widget.groupingItemCount){
              topPosition += widget.groupingItemPadding;
              leftPosition += widget.groupingItemPadding;
              rightPosition += widget.groupingItemPadding;
            }
          }

          bool selectableItem = widget.isExpanded || widget.itemCount == 1;

          return AnimatedPositioned(
            top: topPosition,
            left: leftPosition,
            right: rightPosition,
            duration: widget.animatedGroupDuration,
            child: SizedBox(
              height: widget.itemHeight,
              child: GestureDetector(
                onTap: selectableItem ? (){
                  widget.onSelectItem(index);
                } : null,
                child: widget.itemBuilder(context, index, widget.isExpanded),
              ),
            ),
          );

        }).toList().reversed.toList(),
      ),
    );
  }

  _expandContent() {
    widget.isExpanded = true;
    widget.onContentTrigger(true);
    setState(() {
      widget.isExpanded = true;
    });
  }

}


lib/source/stack_grouping_container.dart
import 'package:flutter/material.dart';
import 'stack_grouping_content.dart';

class StackGroupingContainer extends StatefulWidget {

  bool isExpanded = false;
  int groupingItemCount;
  double itemHeight;
  double groupingItemPadding;
  double expandItemPadding;
  double headerHeight;
  Duration animatedSizeDuration;
  Duration animatedGroupDuration;

  final Widget Function(BuildContext context, int index, bool isExpanded) itemBuilder;
  final Widget Function(BuildContext context, bool isExpanded, VoidCallback onCollapse) headerBuilder;
  final Function(int index) onSelectItem;
  final int itemCount;

  StackGroupingContainer({@required this.itemCount,
    @required this.itemBuilder,
    this.headerBuilder,
    this.isExpanded = false,
    this.groupingItemCount = 3,
    this.itemHeight = 120,
    this.groupingItemPadding = 10,
    this.expandItemPadding = 10,
    this.onSelectItem,
    this.headerHeight = 46,
    this.animatedSizeDuration = const Duration(milliseconds: 350),
    this.animatedGroupDuration = const Duration(milliseconds: 300),
  }) : assert(itemCount != null && itemBuilder != null);

  @override
  _StackGroupingContainerState createState() => _StackGroupingContainerState();
}

class _StackGroupingContainerState extends State<StackGroupingContainer> with SingleTickerProviderStateMixin {

  AnimationController _controller;
  Animation<double> _height;
  Animation<double> _contentTopPosition;
  Animation<double> _headerTopPosition;

  Future<void> triggerContent({bool isExpanded}) async {

    widget.isExpanded = isExpanded;

    try {
      if (isExpanded){
        await _controller.animateTo(1).orCancel;
      }else {
        await _controller.animateTo(0).orCancel;
      }
    } on TickerCanceled {
      // the animation got canceled, probably because it was disposed of
    }

  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  void initState() {

    _controller = AnimationController(vsync: this, duration: widget.animatedSizeDuration);

    _height = Tween<double>(
      begin: _collapseHeight(),
      end: _expandHeight(),
    ).animate(
      CurvedAnimation(
        parent: _controller,
        curve: Curves.ease,
      ),
    );

    _contentTopPosition = Tween<double>(
      begin: 0,
      end: widget.headerHeight,
    ).animate(
      CurvedAnimation(
        parent: _controller,
        curve: Curves.ease,
      ),
    );

    _headerTopPosition = Tween<double>(
      begin: widget.headerHeight,
      end: 0,
    ).animate(
      CurvedAnimation(
        parent: _controller,
        curve: Curves.ease,
      ),
    );

    super.initState();

  }

  double _collapseHeight() {
    if (widget.itemCount == 1){
      return widget.itemHeight + widget.groupingItemPadding;
    }
    double totalHeight = widget.itemHeight;
    for (int index = 0; index < widget.groupingItemCount; index++){
      totalHeight += widget.groupingItemPadding;
    }
    return totalHeight;
  }

  double _expandHeight() {
    if (widget.itemCount == 1){
      return widget.itemHeight + widget.expandItemPadding;
    }
    double totalHeight = 0;
    for (int index = 0; index < widget.itemCount; index++){
      totalHeight += widget.itemHeight;
      totalHeight += widget.expandItemPadding;
    }
    bool hasHeader = widget.headerBuilder != null;
    if (hasHeader){
      totalHeight += widget.headerHeight;
    }
    return totalHeight;
  }

  @override
  Widget build(BuildContext context) {

    return AnimatedBuilder(animation: _controller, builder: (BuildContext context, Widget child){

      bool hasHeader = widget.headerBuilder != null;

      return Container(
        height: _height.value,
        child: Stack(
          children: [
            if (hasHeader)
              Positioned(
                left: 0,
                right: 0,
                top: _headerTopPosition.value,
                height: widget.headerHeight,
                child: widget.headerBuilder(context, widget.isExpanded, (){
                  triggerContent(isExpanded: false);
                }),
              ),
            Positioned(
              left: 0,
              right: 0,
              top: _contentTopPosition.value,
              height: _height.value,
              child: StackGroupingContent(
                isExpanded: widget.isExpanded,
                itemCount: widget.itemCount,
                itemBuilder: widget.itemBuilder,
                itemHeight: widget.itemHeight,
                animatedGroupDuration: widget.animatedGroupDuration,
                expandItemPadding: widget.expandItemPadding,
                groupingItemCount: widget.groupingItemCount,
                groupingItemPadding: widget.groupingItemPadding,
                onSelectItem: widget.onSelectItem,
                onContentTrigger: (isExpanded){
                  triggerContent(isExpanded: isExpanded);
                },
              ),
            ),
          ],
        ),
      );
    },
    );
  }

}


example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:stack_grouping_list/source/stack_grouping_container.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Stack grouping list example',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        backgroundColor: Color(0xFFF5F5F5),
        appBar: CupertinoNavigationBar(
          middle: Text('Stack grouping list example'),
        ),
        body: SafeArea(
          child: ListView(
            padding: EdgeInsets.all(15),
            children: [
              _firstSection(),
              _secondSection(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _firstSection() {
    return StackGroupingContainer(
      itemBuilder: (BuildContext context, int index, bool isExpanded){
        List<BoxShadow> boxShadow = [BoxShadow(blurRadius: 6, color: Colors.black.withOpacity(0.04), offset: Offset(0, 3),)];
        // Prevent shadow too dark
        if (!isExpanded){
          if (index > 3 ){
            boxShadow = null;
          }
        }
        return Container(
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(10),
            color: Colors.white,
            boxShadow: boxShadow,
          ),
          padding: EdgeInsets.all(10),
          child: Column(
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.start,
                children: [
                  Icon(Icons.notifications),
                  SizedBox(width: 10,),
                  Text('MESSENGER'),
                  Spacer(),
                  Text('1m ago'),
                ],
              ),
              SizedBox(height: 4,),
              Expanded(
                child: Align(
                  alignment: Alignment.centerLeft,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('John Wick',
                      ),
                      SizedBox(height: 4,),
                      Text('Everything\'s got a price.',
                      ),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      itemCount: 3,
      headerBuilder: (BuildContext context, bool isExpanded, VoidCallback onDismiss){
        return Padding(
          padding: const EdgeInsets.only(bottom: 12),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Align(
                alignment: Alignment.bottomLeft,
                child: Text('Messenger',
                  style: TextStyle(
                    fontSize: 24,
                  ),
                ),
              ),
              Spacer(),
              Align(
                alignment: Alignment.bottomRight,
                child: GestureDetector(
                  onTap: onDismiss,
                  child: Row(
                    children: [
                      Text('Show less',
                        style: TextStyle(
                          fontSize: 17,
                        ),
                      ),
                      Icon(Icons.keyboard_arrow_up),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      groupingItemCount: 3,
      itemHeight: 90,
      headerHeight: 46,
      expandItemPadding: 15,
      onSelectItem: (index){
        print("select item $index");
      },
    );
  }

  Widget _secondSection() {
    return StackGroupingContainer(
      itemBuilder: (BuildContext context, int index, bool isExpanded){
        List<BoxShadow> boxShadow = [BoxShadow(blurRadius: 6, color: Colors.black.withOpacity(0.04), offset: Offset(0, 3),)];
        // Prevent shadow too dark
        if (!isExpanded){
          if (index > 3 ){
            boxShadow = null;
          }
        }
        return Container(
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(10),
            color: Colors.white,
            boxShadow: boxShadow,
          ),
          padding: EdgeInsets.all(10),
          child: Column(
            children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.start,
                children: [
                  Icon(Icons.notifications),
                  SizedBox(width: 10,),
                  Text('FACEBOOK'),
                  Spacer(),
                  Text('1m ago'),
                ],
              ),
              SizedBox(height: 4,),
              Expanded(
                child: Align(
                  alignment: Alignment.centerLeft,
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('John Wick mentioned you in a post.',),
                      SizedBox(height: 4,),
                      Text('I’m Thinking I’m Back.',),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      itemCount: 5,
      headerBuilder: (BuildContext context, bool isExpanded, VoidCallback onDismiss){
        return Padding(
          padding: const EdgeInsets.only(bottom: 12),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Align(
                alignment: Alignment.bottomLeft,
                child: Text('Facebook',
                  style: TextStyle(
                    fontSize: 24,
                  ),
                ),
              ),
              Spacer(),
              Align(
                alignment: Alignment.bottomRight,
                child: GestureDetector(
                  onTap: onDismiss,
                  child: Row(
                    children: [
                      Text('Show less',
                        style: TextStyle(
                          fontSize: 17,
                        ),
                      ),
                      Icon(Icons.keyboard_arrow_up),
                    ],
                  ),
                ),
              ),
            ],
          ),
        );
      },
      groupingItemCount: 3,
      itemHeight: 90,
      headerHeight: 46,
      expandItemPadding: 15,
      onSelectItem: (index){
        print("select item $index");
      },
    );
  }

}
